macaroonbakery/bakery.py

```diff
@@ -0,0 +1,235 @@
+# Copyright 2016 Canonical Ltd.
+# Licensed under the AGPLv3, see LICENCE file for details.
+
+import base64
+import json
+import requests
+import webbrowser
+
+import macaroons
+
+
+def deserialize(json_macaroon):
+    '''Deserialize a JSON macaroon into a macaroon object.
+
+    @param the JSON macaroon to deserialize.
+    @return the deserialized macaroon.
+    '''
+    return macaroons.deserialize(to_base64(json_macaroon))
+
+
+def to_base64(json_macaroon):
+    '''Base 64 encode a json macaroon
+
+    @param the JSON macaroon to encode.
+    @return the encoded macaroon.
+    '''
+    return base64.urlsafe_b64encode(json.dumps(json_macaroon))
+
+
+def add_padding(s):
+    '''Add padding to base64 encoded string
+    libmacaroon does not give padded base64 string from serialization
+
+    @param string s to be padded
+    @return a padded string
+    '''
+    return s + '=' * ((4 - (len(s) % 4)) % 4)
+
+
+def serialize_macaroon_string(macaroon):
+    '''Serialize macaroon to string
+
+    @param macaroon to be serialized
+    @return a string serialization form of the macaroon
+    '''
+    return base64.urlsafe_b64decode(
+        add_padding(macaroon.serialize_json()))
+
+
+class DischargeError(Exception):
+    pass
+
+
+class Bakery(object):
+    ''' Bakery is a wrapper around requests that ensures
+    that macaroons headers are sent to the server if present or gather the
+    discharged macaroons automatically before sending the request.
+    '''
+    def __init__(self, initial_macaroons=None, discharge_token=None,
+                 timeout=30):
+        '''
+        @param initial_macaroons initial macaroon to be used in headers.
+        @param discharge_token initial discharge token.
+        @param timeout timeout for wait url.
+        '''
+        self.macaroons = initial_macaroons
+        self.discharge_token = discharge_token
+        self.timeout = timeout
+
+    def get(self, url, **kwargs):
+        '''Act as the same than requests.get, same arguments,
+        same returned value, same exception raised except DischargeError
+        when something wrong happened when discharging.
+        If macaroons are present, the headers will be augmented with it.
+        '''
+        if self.macaroons is not None:
```

Comment:
This can be expressed as
```python
if self.macaroons are not None:
    headers = kwargs.setdefault('headers', {})
    headers.update({
        'Bakery-Protocol-Version': 1, 
        'Macaroons': self.macaroons,
    })
```